Introduced logger context

sruke opened the pull request:
Copied over logging properties from CallContext class to newly created LoggerContext class. 

Background -
Properties primarily used for logging are currently part of CallContext class belonging to Microsoft.IdentityModel.Tokens namespace. 
As part of the effort to adopt ILogger I'm also making some improvements to the existing logic. 
This PR's goal is to create a LoggerContext class where all logging properties will reside and CallContext will inherit LoggerContext. This way logging properties will reside in Microsoft.IdentityModel.Logging namespace.

RojaEnnam committed:
Add 'cty' claim to JWE header

sruke committed:
Introduced logger context

sruke committed:
Introduced logger context

sruke commented on src/Microsoft.IdentityModel.Logging/LoggerContext.cs:
```diff
@@ -0,0 +1,81 @@
+        /// <summary>
+        /// Gets or sets a string that helps with setting breakpoints when debugging.
+        /// </summary>
+        public string Id { get; set; } = string.Empty;
```
public string Id will replace an 'internal' property with the same name in SAL's S2SContext class. In the next SAL PR I'll make sure to remove the internal property to avoid compiler getting confused.

GeoK replied:
Why are we exposing this property if it's used only internally?

brentschmaltz replied:
@GeoK It is internal in SAL, but we would like it to be public to aid in debugging.

GeoK commented on src/Microsoft.IdentityModel.Logging/LoggerContext.cs:
```diff
@@ -0,0 +1,81 @@
+        /// <summary>
+        /// Gets or sets an <see cref="IDictionary{String, Object}"/> that enables custom extensibility scenarios.
+        /// </summary>
+        public IDictionary<string, object> PropertyBag { get; set; }
```
This property technically belong more to Call than Logging context.

sruke replied:
Brent's rationale for having PropertyBag under LoggerContext rather than CallContext is that Microsoft.IdentityModel.Logging namespace is more widely used compared to Microsoft.IdentityModel.Tokens which makes creating a LoggerContext object and passing it through the call stack easier.

TimHannMSFT commented on src/Microsoft.IdentityModel.Logging/LoggerContext.cs:
```diff
@@ -0,0 +1,81 @@
+    /// <summary>
+    /// An opaque context used to store work when working with authentication artifacts.
+    /// </summary>
+    public class LoggerContext
```
>LoggerContext [](http://example.com/codeflow?start=17&length=13)

So is it the case that CallContext : LoggerContext instead of CallContext : ILoggerContext mostly from an extensibility standpoint? e.g. If LoggerContext ever changes it's nice to have a class so we don't have to make a whole new interface without it being breaking?

It seems like logically this is an interface, it's saying this is a contract we'd expect such objects to behave under, as opposed to saying CallContexts are a type of LoggingContext.

I have no strong objections but if that's the case we might want to add some comments explaining the rationale.

sruke replied:
Extensibility is one reason. Another reason is this class will be used mostly internally. Right now, I don't see a reason why a customer would want to implement a LoggerContext interface and customize it. Brent wants to see a context object widely available in Wilson to aid with debugging.

TimHannMSFT commented on src/Microsoft.IdentityModel.Logging/LoggerContext.cs:
```diff
@@ -0,0 +1,81 @@
+namespace Microsoft.IdentityModel.Logging
+{
+    /// <summary>
+    /// An opaque context used to store work when working with authentication artifacts.
```
>An opaque context used to store work when working with authentication artifacts [](http://example.com/codeflow?start=8&length=79)

Update here. I don't think the "opaque context" is even true of the CallContext but may as well have this be more accurate on the new class.

ciaozhang committed:
Configuration validator (#1825)

Introduce an ConfigurationValidator class that can be used to apply validation rules to a retrieved configuration.

Commit:
Simplify strings comparison with Ordinal option

brentschmaltz committed:
update version

brentschmaltz committed:
update version for next release

RojaEnnam committed:
Address comments

RojaEnnam committed:
Fix tests (#1838)

RojaEnnam committed:
Fix DEF test (#1839)

sruke committed:
Added custom logger interface (#1823)

jennyf19 commented:
Will this be merged and does it have a corresponding issue? @sruke

jennyf19 committed:
log cert thumbprint (#1820)

* log cert thumbprint

* PR feedback

sruke committed:
Introduced custom log level enum to remove dependency on System.Diagnostics.Tracing.EventLevel in IIdentityLogger (#1843)

brentschmaltz committed:
update version to 6.18.1

brentschmaltz committed:
Support for ECDH-ES

sruke committed:
Introducing IdentityLoggerAdapter for Identity Libraries integrating (#1862)

sruke committed:
Fix IIdentityLogger Unit Test (#1863)

brentschmaltz committed:
moved all settings to strongnamebypass and fixed error with WOW6432Node

fixed typos

TimHannMSFT committed:
Create thin Abstractions package and add Telemetry interface (#1813)

* Add to thin Abstractions package and add Telemetry interface

* Created ITelemetryClient interface
* Created NullTelemetryClient implmentation which defaults to no-ops
* Add unit tests for abstractions package

* Addressed PR Feedback
* Updates to IDictionary surface
* Updates to avoid dictionary copy
* Updates to show an example of extending the base type for TelemetryEventDetails to include dataClassification

brentschmaltz committed:
updated version to 6.19.1

dannybtsai committed:
fixed the issue the compaction actions are queued but potentially never got started (#1871)

TimHannMSFT committed:
[Bugfix] fix casing for parameter "DateTimeProperties" => "dateTimeProperties" (#1870)

- parameters should always start with a lowercase letter then camel case to match existing code
  and expectations for .NET libraries.

dannybtsai committed:
fixed a test failure because the event queue task needs more time to … (#1874)

* fixed a test failure because the event queue task needs more time to start

* replaced possibly longer-than-needed sleep with more precise checks (for completion of the cache compaction action)

sruke committed:
Merge branch 'sruthi/LoggerContext' of https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet into sruthi/LoggerContext

sruke committed:
Addressed feedback. Added test.

sruke commented:
Rebase is not straightforward as its a very old PR. Will open a new PR.
